monkestation/code/modules/hydroponics/machines/genesischamber.dm

```diff
@@ -0,0 +1,255 @@
+		update_appearance()
+
+/obj/machinery/genesis_chamber/RefreshParts()
+	. = ..()
+	max_capacity = initial(max_capacity)
+	for(var/datum/stock_part/matter_bin/bin in component_parts)
+		max_capacity += max((bin.tier * 5) - 5, 0)
+
+	max_cooldown = initial(max_cooldown)
+	for(var/datum/stock_part/micro_laser/laser in component_parts)
+		max_cooldown -= max((laser.tier - 1) * (1 MINUTES), 0 MINUTES)
+
+	// Reset the timer if the machine is active
+	if(on)
+		if(seed_timer_id)
+			deltimer(seed_timer_id) // Stop the old timer
+		generation_start_time = world.time // Reset the start time
+		seed_timer_id = addtimer(CALLBACK(src, PROC_REF(generate_seed)), max_cooldown, TIMER_STOPPABLE | TIMER_UNIQUE | TIMER_DELETE_ME)
+
+/obj/machinery/genesis_chamber/attack_paw(mob/user, list/modifiers)
+	return attack_hand(user, modifiers)
+
+/obj/machinery/genesis_chamber/attack_hand(mob/living/user, list/modifiers)
+	. = ..()
+
+	if(obj_flags & EMAGGED)
+		return
+
+	if(panel_open)
+		to_chat(user, span_warning("You can't use the [src] while the maintenance panel is open!"))
+		return
+
+	if(capacity < 1)
+		calculate_time_remaining()
+		to_chat(user, span_warning("You need to wait [DisplayTimeText(time_remaining)] until at least one seed is produced."))
+		return
+
+	if(QDELETED(user) || QDELETED(src) || !user.can_perform_action(src, FORBID_TELEKINESIS_REACH | ALLOW_RESTING))
+		return
+
+	for (var/i in 1 to capacity)
+		if(capacity <= 0)
+			break
+		new /obj/item/seeds/random(drop_location())
+
+	to_chat(user, span_notice("You collect [capacity] strange seeds from the chamber."))
+	capacity = 0
+	update_appearance()
+
+/obj/machinery/genesis_chamber/proc/calculate_time_remaining()
+	if(!on || !seed_timer_id)
+		time_remaining = max_cooldown
+		return
+
+	var/elapsed_time = world.time - generation_start_time
+	time_remaining = max(0, max_cooldown - elapsed_time)
+
+/obj/machinery/genesis_chamber/attack_hand_secondary(mob/living/user, list/modifiers)
+	. = ..()
+	if(. == SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN)
+		return
+
+	if(obj_flags & EMAGGED)
+		return
+
+	if(panel_open)
+		return
+
+	on = !on
+
+	if(on)
+		to_chat(user, span_notice("You turn on [src]."))
+		playsound(loc, 'sound/machines/chime.ogg', 40, TRUE)
+		say("Activating... Seeds will be produced in [DisplayTimeText(max_cooldown)].")
+		START_PROCESSING(SSprocessing, src)
+		generation_start_time = world.time
+		seed_timer_id = addtimer(CALLBACK(src, PROC_REF(generate_seed)), max_cooldown, TIMER_STOPPABLE | TIMER_UNIQUE | TIMER_DELETE_ME)
+	else
+		to_chat(user, span_notice("You turn off [src]."))
+		playsound(loc, 'sound/machines/click.ogg', 15, TRUE, -3)
+		say("Deactivating...")
+		STOP_PROCESSING(SSprocessing, src)
+		if(seed_timer_id)
+			deltimer(seed_timer_id)
+			seed_timer_id = null
+		time_remaining = max_cooldown
+
+	update_appearance()
+	return SECONDARY_ATTACK_CANCEL_ATTACK_CHAIN
+
+/obj/machinery/genesis_chamber/process()
+	if(!on)
+		return PROCESS_KILL
+
+	calculate_time_remaining()
```

Comment:
PLEASE PLEASE PLEASE do not use `process()` to calculate time remaining.

Use cooldowns (see [code/__DEFINES/cooldowns.dm](https://github.com/Monkestation/Monkestation2.0/blob/master/code/__DEFINES/cooldowns.dm)), or if you can't use those for some reason, make sure to use `DELTA_WORLD_TIME` to ensure lagspikes don't cause issues.